File: cs/Problems/Day25.cs
namespace AOC2023.Problems;

public class Day25 : IProblem<int>
{
    public int Solve(ReadOnlySpan<char> input) => ProductOfComponentGroups(input);

    private int ProductOfComponentGroups(ReadOnlySpan<char> input)
    {
        Span<Range> lineRanges = stackalloc Range[1275];
        Span<Range> splitRanges = stackalloc Range[10];
        int lines = input.Split(lineRanges, InputReader.NewLine);

        var edges = new List<(int From, int To)>(lines * 3);

        for (int i = 0; i < lines; i++)
        {
            var line = input[lineRanges[i]];
            int connectionCount = line.Split(splitRanges, " ");

            for (int j = 1; j < connectionCount; j++)
                edges.Add((string.GetHashCode(line[..3]), string.GetHashCode(line[splitRanges[j]])));
        }

        var mergeAlgorithm = new KargerAlgorithm(edges);
        var (cutSize, group1Size, group2Size) = (0, 0, 0);

        while (cutSize != 3)
            (cutSize, group1Size, group2Size) = mergeAlgorithm.FindMinimumCut();

        return group1Size * group2Size;
    }

    // Found this algorithm somewhere else.
    readonly struct KargerAlgorithm
    {
        private readonly int initialVerticesCount;
        private readonly List<(int From, int To)> initialEdges;
        private readonly Random rng = new(Seed: 72);

        private readonly SortedList<int, List<int>> merged = [];
        private readonly List<(int From, int To)> cache1 = [];
        private readonly List<(int From, int To)> cache2 = [];

        public KargerAlgorithm(List<(int From, int To)> edges)
        {
            initialEdges = edges;

            var uniqueVertices = new HashSet<int>();
            foreach (var (from, to) in edges)
            {
                initialVerticesCount += uniqueVertices.Add(from) ? 1 : 0;
                initialVerticesCount += uniqueVertices.Add(to) ? 1 : 0;
            }
        }

        public readonly (int CutSize, int Group1Count, int Group2Count) FindMinimumCut()
        {
            var mergedEdges = initialEdges;
            var mergedVerticesCount = initialVerticesCount;

            // Clears and reuses list instances. Not amazing but otherwise this becomes a memory hog.
            merged.Clear();
            cache1.Clear();
            cache2.Clear();

            while (mergedVerticesCount > 2)
            {
                var (from, to) = mergedEdges[rng.Next(0, mergedEdges.Count)];

                if (!merged.ContainsKey(from))
                    merged[from] = [];

                merged[from].Add(to);

                if (merged.TryGetValue(to, out List<int>? connections))
                {
                    merged[from].AddRange(connections);
                    merged.Remove(to);
                }

                var newEdges = cache1 == mergedEdges ? cache2 : cache1; // Prevents creating a new list on each iter.
                newEdges.Clear();

                foreach (var edge in mergedEdges)
                {
                    if (edge.To == to)
                    {
                        if (edge.From != from)
                            newEdges.Add((edge.From, from));
                    }
                    else if (edge.From == to)
                    {
                        if (from != edge.To)
                            newEdges.Add((from, edge.To));
                    }
                    else
                    {
                        newEdges.Add(edge);
                    }
                }

                (mergedEdges, _) = (newEdges, mergedEdges);
                mergedVerticesCount--;
            }

            var vertexGroups = merged.Values;
            return (mergedEdges.Count, vertexGroups[0].Count + 1, vertexGroups[^1].Count + 1);
        }
    }
}

File: go/problems/day25.go
package problems

import (
	"math/rand"
	"strings"
)

type Day25 struct{}

func (d Day25) Solve(input string) int {
	return productOfComponentGroups(input)
}

func productOfComponentGroups(input string) int {
	lines := strings.Split(input, NewLine)
	edges := make([]Edge, 0, len(lines)*3)

	for _, line := range lines {
		connections := strings.Split(line, " ")

		fromHash := connections[0]
		for _, connection := range connections[1:] {
			toHash := connection
			edges = append(edges, Edge{From: fromHash, To: toHash})
		}
	}

	mergeAlgorithm := NewKargerAlgorithm(edges)
	cutSize, group1Size, group2Size := 0, 0, 0

	for cutSize != 3 {
		cutSize, group1Size, group2Size = mergeAlgorithm.FindMinimumCut()
	}

	return group1Size * group2Size
}

type Edge struct {
	From, To string
}

type KargerAlgorithm struct {
	initialVerticesCount int
	initialEdges         []Edge
	rng                  *rand.Rand
	merged               map[string][]string
}

func NewKargerAlgorithm(edges []Edge) *KargerAlgorithm {
	ka := &KargerAlgorithm{
		initialEdges: edges,
		rng:          rand.New(rand.NewSource(72)),
		merged:       make(map[string][]string),
	}

	uniqueVertices := make(map[string]bool)
	for _, e := range edges {
		if _, ok := uniqueVertices[e.From]; !ok {
			ka.initialVerticesCount++
			uniqueVertices[e.From] = true
		}
		if _, ok := uniqueVertices[e.To]; !ok {
			ka.initialVerticesCount++
			uniqueVertices[e.To] = true
		}
	}
	return ka
}

func (ka *KargerAlgorithm) FindMinimumCut() (cutSize, group1Count, group2Count int) {
	mergedEdges := ka.initialEdges
	mergedVerticesCount := ka.initialVerticesCount
	ka.merged = make(map[string][]string)

	for mergedVerticesCount > 2 && len(mergedEdges) > 0 {
		index := ka.rng.Intn(len(mergedEdges))
		from, to := mergedEdges[index].From, mergedEdges[index].To

		ka.merged[from] = append(ka.merged[from], to)
		if connections, ok := ka.merged[to]; ok {
			ka.merged[from] = append(ka.merged[from], connections...)
			delete(ka.merged, to)
		}

		newEdges := make([]Edge, 0)

		for _, edge := range mergedEdges {
			if edge.To == to {
				if edge.From != from {
					newEdges = append(newEdges, Edge{From: edge.From, To: from})
				}
			} else if edge.From == to {
				if from != edge.To {
					newEdges = append(newEdges, Edge{From: from, To: edge.To})
				}
			} else {
				newEdges = append(newEdges, edge)
			}
		}

		mergedEdges = newEdges
		mergedVerticesCount--
	}

	vertexGroups := make([][]string, 0, len(ka.merged))
	for _, value := range ka.merged {
		vertexGroups = append(vertexGroups, value)
	}
	return len(mergedEdges), len(vertexGroups[0]) + 1, len(vertexGroups[len(vertexGroups)-1]) + 1
}
